File: 0207-course-schedule/0207-course-schedule.py
#detect cycle using DFS
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        adjList = []
        for i in range(numCourses):
            adjList.append([])
        visited = [False]*numCourses
        cycleDetection = False
        for p in prerequisites:
            adjList[p[1]].append(p[0])
        cycleSet = set()
        cycleSet.add(0)
        def DFS(s):
            nonlocal cycleDetection
            cycleSet.add(s)
            for node in adjList[s]:
                if visited[node] == False:
                    visited[node] = True
                    DFS(node)
                elif node in cycleSet:
                    cycleDetection = True
                    return
            cycleSet.remove(s)
            
        
        for i in range(numCourses):
            if visited[i] == False:
                visited[i] = True
                DFS(i)
        if cycleDetection == True:
            return False
        return True
        
        

File: 0207-course-schedule/0207-course-schedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        // Step 1: Create an indegree array to track number of incoming edges for each course
        int[] indegree = new int[numCourses];
        // Step 2: Create an adjacency list to represent the graph
        List<List<Integer>> adj = new ArrayList<>(numCourses);
        // Initialize the adjacency list with empty lists for each course
        for (int i = 0; i < numCourses; i++) {
            adj.add(new ArrayList<>());
        }
        // Step 3: Build the graph from the prerequisites
        for (int[] prerequisite : prerequisites) {
            int course = prerequisite[0];
            int dependency = prerequisite[1];
            // dependency -> course
            adj.get(dependency).add(course); // Add edge to adjacency list
            indegree[course]++;              // Increase indegree of the course
        }
        // Step 4: Initialize the queue with all courses having 0 indegree (no prerequisites)
        Queue<Integer> queue = new LinkedList<>();
        for (int i = 0; i < numCourses; i++) {
            if (indegree[i] == 0) {
                queue.offer(i);
            }
        }
        // Step 5: Process nodes using BFS (Kahn's Algorithm)
        int nodesVisited = 0;
        while (!queue.isEmpty()) {
            int node = queue.poll(); // Remove a node with indegree 0
            nodesVisited++;          // Count the course as completed
            // Visit all the neighbors (courses that depend on current course)
            for (int neighbor : adj.get(node)) {
                indegree[neighbor]--; // Remove the edge
                // If neighbor now has no prerequisites, add it to the queue
                if (indegree[neighbor] == 0) {
                    queue.offer(neighbor);
                }
            }
        }
        // Step 6: If we've visited all courses, we can finish all
        return nodesVisited == numCourses;
    }
}
